lib/std/os.zig

```diff
@@ -4981,22 +4987,18 @@ pub const CopyFileRangeError = error{
 pub fn copy_file_range(fd_in: fd_t, off_in: u64, fd_out: fd_t, off_out: u64, len: usize, flags: u32) CopyFileRangeError!usize {
     const use_c = std.c.versionCheck(.{ .major = 2, .minor = 27, .patch = 0 }).ok;
 
-    // TODO support for other systems than linux
-    const try_syscall = comptime std.Target.current.os.isAtLeast(.linux, .{ .major = 4, .minor = 5 }) != false;
-
-    if (use_c or try_syscall) {
+    if (std.Target.current.os.tag == .linux and
+        (use_c or has_copy_file_range_syscall.get()))
+    {
         const sys = if (use_c) std.c else linux;
 
         var off_in_copy = @bitCast(i64, off_in);
         var off_out_copy = @bitCast(i64, off_out);
 
         const rc = sys.copy_file_range(fd_in, &off_in_copy, fd_out, &off_out_copy, len, flags);
-
-        // TODO avoid wasting a syscall every time if kernel is too old and returns ENOSYS https://github.com/ziglang/zig/issues/1018
-
         switch (sys.getErrno(rc)) {
             0 => return @intCast(usize, rc),
-            EBADF => unreachable,
+            EBADF => return error.InvalidFileDescriptor,
```

Comment:
The man page for copy_file_range says that this means one of 3 possible things:

       EBADF  One or more file descriptors are not valid.

       EBADF  fd_in is not open for reading; or fd_out is not open  for  writ‐
              ing.

       EBADF  The  O_APPEND  flag  is  set  for the open file description (see
              open(2)) referred to by the file descriptor fd_out.

=> unreachable is appropriate for the first case, since it would be a possible Use-After-Free to try to handle this; it is programmer error to ever encounter it.

However, those second 2 possibilities are viable error conditions. In std.os.write we have:

            EBADF => return error.NotOpenForWriting, // can be a race condition.

and in std.os.read:

            EBADF => return error.NotOpenForReading, // Can be a race condition.

I suggest this change:

```suggestion
            EBADF => return error.FilesOpenedWithWrongFlags,
```

With these doc comments added in the error set:

```zig
/// `fd_in` is not open for reading; or `fd_out` is not open  for  writing; or
/// the  `O_APPEND`  flag  is  set  for `fd_out`.
```

Reply:
`EBADF => unreachable` means no-op in `releasefast` mode and IMO that's dangerous if you don't have complete control over the fd. And even then you never know what weird trick the kernel may pull on you...